### PR TITLE
Build TypeScript plugin with TypeScript 4

### PR DESCRIPTION
**Pull request by benmccann**
## Rollup Plugin Name: all

This PR contains:

- [ ] bugfix
- [ ] feature
- [X] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [X] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [X] no

I don't expect this would cause folks issues in upgrading. The only code changes I had to make were slight and internal to the plugins, so I don't think the external types will change

### Description

There are a number of open issues with the TypeScript plugin. I was poking around to see what would be required to fix some of them. Being on an older version of TypeScript makes things harder. E.g. it seems like there might be a bug in one of the TypeScript methods (https://github.com/rollup/plugins/issues/287#issuecomment-674552597), but I can't report issues to the TypeScript team unless using the latest version of TypeScript. I also was trying to use an API that I saw in `tsc` and appeared to be exported, but VS Code is telling me it doesn't exist and I wonder if being on an older version of TypeScript might be why.

### CODE REVIEW

**Comment by shellscape** [comment]
Appreciate the effort here. For PRs that will affect more than one plugin, the changes should really be discussed ahead of time. Looks like I'm going to need to update the contributing document to cover this. PRs that affect multiple plugins should be rare and require a special commit history and a regular merge commit versus squash commit - this is to facilitate an independent plugin publishing process (versus babel where they publish the world for each new version).

We should hold off on this right now. If you'd like to update just the typescript plugin, roll the other plugins back, and repurpose this branch and PR for only that plugin, then I can get behind that. I leave that choice up to you.

**Comment by benmccann** [comment]
Ok. I will update this PR an open an issue to discuss the process of updating all plugins

**Comment by shellscape** [comment]
@NotWoods this was rebased to only affect the Typescript plugin, so it'll need your approval.

File: packages/typescript/package.json
Comment (NotWoods) [inline packages/typescript/package.json:62]: Let's update to the latest version (4.1.2)

Comment (benmccann) [reply packages/typescript/package.json]: done (though the `^` prefix means get the latest 4.x version in either case, so they are equivalent)

**Review by NotWoods — APPROVED** [review]

**Pull request merged**
